src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/DefaultWorkspaceSemanticTokensRefreshTrigger.cs

```diff
@@ -0,0 +1,48 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer
+{
+    /// <summary>
+    /// Sends a 'workspace\semanticTokens\refresh' request each time the project changes.
+    /// </summary>
+    internal class DefaultWorkspaceSemanticTokensRefreshTrigger : ProjectSnapshotChangeTrigger
+    {
+        private readonly WorkspaceSemanticTokensRefreshPublisher _workspaceChangedPublisher;
+        private ProjectSnapshotManagerBase? _projectSnapshotManager;
+
+        internal DefaultWorkspaceSemanticTokensRefreshTrigger(WorkspaceSemanticTokensRefreshPublisher workspaceSemanticTokensRefreshPublisher)
+        {
+            _workspaceChangedPublisher = workspaceSemanticTokensRefreshPublisher;
+        }
+
+        public override void Initialize(ProjectSnapshotManagerBase projectManager)
+        {
+            _projectSnapshotManager = projectManager;
+
+            _projectSnapshotManager.Changed += ProjectSnapshotManager_Changed;
+
+            var errorReporter = GetErrorReporter();
+            _workspaceChangedPublisher.Initialize(errorReporter);
+        }
+
+        internal virtual ErrorReporter GetErrorReporter()
+        {
+            return _projectSnapshotManager!.Workspace.Services.GetRequiredService<ErrorReporter>();
```

Comment:
Lets register an error reporter on the default service collection so we don't need to pull from the workspace services. We've been trying to move away from using those